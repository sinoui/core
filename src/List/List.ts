import styled, { css } from 'styled-components';
import classNames from 'classnames';

/**
 * List  列表根组件
 */

export interface Props {
  /**
   * 设置为true，表示是嵌入式列表
   */
  insert?: boolean;
  /**
   * 设置为true,表示禁用 ripple 效果
   */
  disabledRipple?: boolean;
  /**
   * 设置 List 的子元素的左侧内边距
   */
  paddingLeft?: number;
  /**
   * 设置为true,表示为密集模式
   */
  dense?: boolean;
  /**
   * 自定义class名称
   */
  className?: string;
}

const ListStyle = css<Props>`
  list-style: none;
  margin: 0;
  padding: 0;
  box-sizing: border-box;

  & > *.sinoui-list-item {
    ${({ paddingLeft }) =>
      paddingLeft !== undefined && `padding-left: ${paddingLeft}px`}
  }

  & > *.sinoui-list-item--insert {
  }

  & > *.sinoui-list-item--disabled-ripple {
  }
`;

const List = styled.ul.attrs((props: Props) => ({
  className: classNames('sinoui-list', {
    'sinoui-list--insert': props.insert,
    'sinoui-list--disabled-ripple': props.disabledRipple,
    'sinoui-list--dense': props.dense,
  }),
}))<Props>`
  ${ListStyle}
`;

export default List;
