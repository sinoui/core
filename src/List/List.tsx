import React, { useEffect } from 'react';
import styled, { css } from 'styled-components';
import classNames from 'classnames';
import OverriableComponent from '../OverridableComponent';
import useMultiRefs from '../utils/useMultiRefs';

/**
 * List  列表根组件
 */

export interface Props {
  /**
   * 设置为true，表示是嵌入式列表
   */
  insert?: boolean;
  /**
   * 设置为true,表示禁用 ripple 效果
   */
  disabledRipple?: boolean;
  /**
   * 设置 List 的子元素的左侧内边距
   */
  paddingLeft?: number;
  /**
   * 设置为true,表示为密集模式
   */
  dense?: boolean;
  /**
   * 自定义class名称
   */
  className?: string;
}

const ListStyle = css<Props>`
  list-style: none;
  margin: 0;
  padding: 8px 0;
  box-sizing: border-box;

  &.sinoui-list--insert > .sinoui-list-item {
    padding-left: 32px;
  }

  & > .sinoui-list-item {
    ${({ paddingLeft }) =>
      paddingLeft !== undefined && `padding-left: ${paddingLeft}px`}
  }
`;

const StyledList = styled.ul.attrs((props: Props) => ({
  className: classNames('sinoui-list', {
    'sinoui-list--insert': props.insert,
    'sinoui-list--disabled-ripple': props.disabledRipple,
    'sinoui-list--dense': props.dense,
  }),
}))<Props>`
  ${ListStyle}
`;

// 将nodeList转换为数组
const nodeListToArray = (listRef: HTMLElement | null) => {
  const arr = [];
  const anchors = listRef?.querySelectorAll('.sinoui-list-item') || [];
  anchors.forEach((item: any) => {
    arr.push(item);
  });

  return arr;
};

const List: OverriableComponent<Props, 'ul'> = React.forwardRef<
  HTMLUListElement,
  Props
>((props, ref) => {
  const listRef = React.createRef<HTMLElement>();
  const multiRef = useMultiRefs(listRef, ref);

  useEffect(() => {
    const listItemsLen =
      listRef.current?.querySelectorAll('.sinoui-list-item').length || 0;
    document.onkeydown = function(e) {
      let focusIndex = nodeListToArray(listRef.current).findIndex(
        (item) => item === document.activeElement,
      );
      if (listItemsLen > 0) {
        switch ((e || window.event).keyCode) {
          case 38:
            focusIndex = focusIndex > 0 ? focusIndex - 1 : 0;
            break;
          case 40:
            focusIndex =
              focusIndex < listItemsLen - 1 ? focusIndex + 1 : listItemsLen - 1;
            break;
          default:
            break;
        }
        (listRef.current?.querySelectorAll('.sinoui-list-item')[
          focusIndex
        ] as HTMLInputElement).focus();
      }
    };
    return () => {
      document.onkeydown = null;
    };
  }, [listRef]);
  const { children, ...rest } = props;
  const renderChildren = () =>
    React.Children.map(
      children as React.ReactElement<any>,
      (child: React.ReactElement<any>) => {
        return React.cloneElement(React.Children.only(child), {
          insert: child.props.insert || props.insert,
          dense: child.props.dense || props.dense,
          paddingLeft: child.props.paddingLeft || props.paddingLeft,
          disabledRipple: child.props.disabledRipple || props.disabledRipple,
        });
      },
    );

  return (
    <StyledList {...rest} ref={multiRef}>
      {renderChildren()}
    </StyledList>
  );
});

export default List;
