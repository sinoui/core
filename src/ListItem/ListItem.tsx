import React from 'react';
import styled, { css } from 'styled-components';
import classNames from 'classnames';
import { opacify } from 'polished';
import { useRipple } from '@sinoui/ripple';
import OverriableComponent from '../OverridableComponent';
import useMultiRefs from '../utils/useMultiRefs';

/**
 * ListItem  列表项组件
 */

export interface ListItemProps {
  /**
   * 设置为true，表示是嵌入样式
   */
  insert?: boolean;
  /**
   * 设置为true,表示禁用 ripple 效果
   */
  disabledRipple?: boolean;
  /**
   * 左侧内边距
   */
  paddingLeft?: number;
  /**
   * 设置为true,表示为密集模式
   */
  dense?: boolean;
  /**
   * 若设置为true，则为不可用状态，不可访问
   */
  disabled?: boolean;
  /**
   * 若设置为true，则为选中状态
   */
  selected?: boolean;
  /**
   * ListItem子元素垂直方向的布局方式
   */
  alignItems?: 'flex-start' | 'flex-end' | 'center';
  /**
   * 单击回调函数
   */
  onClick?: (e: React.MouseEvent, value?: any) => void;
  /**
   * 双击回调函数
   */
  onDoubleClick?: (e: React.MouseEvent, value?: any) => void;
  /**
   * 自定义class名称
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
}

/**
 * 选中样式
 */
const selectedCss = css<ListItemProps>`
  background-color: ${(props) =>
    opacify(
      props.theme.palette.action.selectedOpacity - 1,
      props.theme.palette.primary.main,
    )};
`;

const focusCss = css<ListItemProps>`
  &:focus::before {
    background-color: ${({ theme, selected }) =>
      !selected
        ? opacify(-0.88, theme.palette.type === 'light' ? '#000' : '#fff')
        : opacify(-0.8, theme.palette.primary.main)};
  }
`;

const hoverCss = css<ListItemProps>`
  &:hover {
    ::before {
      background-color: ${({ theme, selected }) =>
        !selected
          ? theme.palette.action.hover
          : opacify(-0.88, theme.palette.primary.main)};
    }
    @media (hover: none) {
      &::before {
        background-color: transparent;
      }
    }
  }
`;

const disabledCss = css`
  color: ${({ theme }) => theme.palette.text.disabled};
  cursor: default;
  pointer-events: none;
`;

const ListItemStyle = css<ListItemProps>`
  position: relative;
  display: flex;
  align-items: ${({ alignItems = 'center' }) => alignItems};
  margin: 0;
  box-sizing: border-box;
  padding: 12px 16px;
  background-color: ${(props) => props.theme.palette.background.paper};
  outline: none;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  ${(props) => props.selected && selectedCss};
  ${(props) => !props.disabled && hoverCss};
  ${(props) => !props.disabled && focusCss};
  ${(props) => props.disabled && disabledCss};
  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 0;
  }

  ${({ paddingLeft }) =>
    paddingLeft !== undefined && `padding-left: ${paddingLeft}px`}

  ${({ insert }) => insert && `padding-left: 32px`}
`;

const Wrapper = styled.li.attrs((props: ListItemProps) => ({
  tabIndex: '0',
  className: classNames('sinoui-list-item', {
    'sinoui-list-item--insert': props.insert,
    'sinoui-list-item--disabled-ripple': props.disabledRipple,
    'sinoui-list-item--dense': props.dense,
  }),
}))<ListItemProps>`
  ${ListItemStyle}
`;

const ListItem: OverriableComponent<ListItemProps, 'li'> = React.forwardRef<
  HTMLLIElement,
  ListItemProps
>((props, ref) => {
  const { disabledRipple, disabled } = props;
  const rippleConfig = disabledRipple
    ? { enableKeyboardFocus: false, disabled: true }
    : { enableKeyboardFocus: false, disabled };
  const rippleRef = useRipple<HTMLLIElement>(rippleConfig);
  const listItemRef = useMultiRefs(rippleRef, ref);

  return <Wrapper {...props} ref={listItemRef} />;
});
export default ListItem;
