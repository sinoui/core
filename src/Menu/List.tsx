import React from 'react';
import styled, { css } from 'styled-components';

export interface ListProps {
  children: React.ReactNode;
  subheader?: React.ReactElement<any>;
  dense?: boolean;
  disablePadding?: boolean;
  role?: string;
  onBlur?: (event: React.FocusEvent<HTMLUListElement>) => void;
  color?: string;
  textStyle?: React.CSSProperties;
}

const paddingStyle = css`
  padding-top: ${(props) => props.theme.spacing.unit}px;
  padding-bottom: ${(props) => props.theme.spacing.unit}px;
`;

const denseStyle = css`
  padding-top: ${(props) => props.theme.spacing.unit / 2}px;
  padding-bottom: ${(props) => props.theme.spacing.unit / 2}px;
`;

const subheaderStyle = css`
  padding-top: 0px;
`;

const StyleList = styled.ul<ListProps>`
  display: flex;
  flex-direction: column;
  flex: 1 1 auto;
  margin: 0px;
  padding: 0px;
  position: relative;
  list-style: none;
  ${(props) => props.dense && !props.disablePadding && denseStyle};
  ${(props) => !props.disablePadding && paddingStyle};
  ${(props) => props.subheader && subheaderStyle};
`;

export default React.forwardRef<HTMLUListElement, ListProps>(function List(
  props,
  ref,
) {
  const { children, subheader, ...other } = props;

  return (
    <StyleList subheader={subheader} {...other} ref={ref}>
      {subheader}
      {React.Children.map(
        children as any,
        (
          child: React.ReactElement<{
            dense?: boolean;
            color?: string;
            textStyle?: any;
          }>,
        ) => {
          if (React.isValidElement(child)) {
            return React.cloneElement(child, {
              dense: props.dense,
              color: props.color,
              textStyle: props.textStyle,
            });
          }
          return child;
        },
      )}
    </StyleList>
  );
});
