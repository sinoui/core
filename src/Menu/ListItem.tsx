import React from 'react';
import styled, { css } from 'styled-components';
import { withRipple } from '@sinoui/ripple';

export interface ListItemProps {
  button?: boolean;
  children?: React.ReactNode;
  component?: React.ReactType;
  dense?: boolean;
  disabled?: boolean;
  disableGutters?: boolean;
  divider?: boolean;
  selected?: boolean;
  tabIndex?: number;
  textStyle?: React.CSSProperties;
  color?: string;
  value?: any;
  onClick?: (event: React.MouseEvent<any>) => void;
  role?: string;
}

const denseStyle = css`
  padding: 8px 12px;
`;

const gutterStyle = css`
  padding-left: ${(props) => props.theme.spacing.unit * 2}px;
  padding-right: ${(props) => props.theme.spacing.unit * 2}px;
`;

const dividerStyle = css`
  border-bottom: 1px solid ${(props) => props.theme.palette.divider};
`;

const buttonStyle = css`
  transition: ${(props) =>
    props.theme.transitions.create('background-color', {
      duration: props.theme.transitions.duration.shortest,
    })};
  ${(props: ListItemProps) => props.disabled && 'background-color:transparent'};

  &:hover {
    text-decoration: none;
    background-color: ${(props) => props.theme.palette.action.hover};

    @media (hover: none) {
      background-color: transparent;
    }
  }
`;

const Typography = styled.div`
  font-size: ${(props) => props.theme.typography.subtitle1.fontSize}rem;
  color: ${(props) =>
    props.color ? props.color : props.theme.palette.text.primary};
`;

const StyleItem = styled.li<ListItemProps>`
  display: flex;
  justify-content: flex-start;
  align-items: center;
  position: relative;
  text-decoration: none;
  padding-top: 12px;
  padding-bottom: 12px;
  outline: none;
  ${(props) => !props.disableGutters && gutterStyle};
  ${(props) => props.divider && dividerStyle};
  ${(props) => props.disabled && 'opacity: 0.5'};
  ${buttonStyle};
  ${(props) => props.dense && denseStyle};
`;

const StyleBaseButton = withRipple({
  enableKeyboardFocus: false,
})(StyleItem);

/**
 * Menu组件中的列表项组件
 *
 */
function ListItem(props: ListItemProps) {
  const {
    button = true,
    children: childrenProp,
    component: componentProp,
    disabled,
    ...other
  } = props;
  const itemChildren = React.Children.map(childrenProp, (child) => {
    if (typeof child === 'string') {
      return (
        <Typography
          key="menu-title"
          color={props.color}
          style={props.textStyle}
        >
          {child}
        </Typography>
      );
    }
    return child;
  });

  const listItemProps = {
    disabled,
    button,
    component: undefined,
    ...other,
  };
  let ComponentMain = componentProp || StyleItem;

  if (button) {
    ComponentMain = StyleBaseButton;
    listItemProps.component = componentProp as any;
  }

  return <ComponentMain {...listItemProps}>{itemChildren}</ComponentMain>;
}

export default ListItem;
