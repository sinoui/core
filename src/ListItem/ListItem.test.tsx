import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import renderer from 'react-test-renderer';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';
import List from '@sinoui/core/List';
import ListItem from '@sinoui/core/ListItem';

/**
 *  单元测试
 */
describe('List', () => {
  afterEach(cleanup);
  it('自定义class名称', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <List>
          <ListItem data-testid="listItem" className="test">
            item
          </ListItem>
        </List>
      </ThemeProvider>,
    );
    expect(getByTestId('listItem')).toHaveClass('test');
  });

  it('指定不同的根元素', () => {
    const { container } = render(
      <ThemeProvider theme={defaultTheme}>
        <List as="div">
          <ListItem as="div">item</ListItem>
        </List>
      </ThemeProvider>,
    );
    expect(container.firstChild && container.firstChild.nodeName).toBe('DIV');
  });

  it('不同属性值下的样式类', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <List>
          <ListItem data-testid="listItem" insert dense disabledRipple>
            item
          </ListItem>
        </List>
      </ThemeProvider>,
    );
    expect(getByTestId('listItem')).toHaveClass('sinoui-list-item--insert');
    expect(getByTestId('listItem')).toHaveClass('sinoui-list-item--dense');
    expect(getByTestId('listItem')).toHaveClass(
      'sinoui-list-item--disabled-ripple',
    );
    // 内嵌列表项有32px的左边距
    expect(getByTestId('listItem')).toHaveStyle('padding-left: 32px');
  });

  it('ListItem设置72px的左内边距', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <List>
          <ListItem data-testid="listItem" paddingLeft={72}>
            item
          </ListItem>
        </List>
      </ThemeProvider>,
    );
    expect(getByTestId('listItem')).toHaveStyle('padding-left: 72px');
  });

  it('单击事件', () => {
    const testClick = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <List>
          <ListItem data-testid="listItem" onClick={testClick}>
            item
          </ListItem>
        </List>
      </ThemeProvider>,
    );
    act(() => {
      fireEvent.click(getByTestId('listItem'));
    });
    expect(testClick).toHaveBeenCalled();
  });

  it('双击事件', () => {
    const testDoubleClick = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <List>
          <ListItem data-testid="listItem" onDoubleClick={testDoubleClick}>
            item
          </ListItem>
        </List>
      </ThemeProvider>,
    );
    act(() => {
      fireEvent.doubleClick(getByTestId('listItem'));
    });
    expect(testDoubleClick).toHaveBeenCalled();
  });
});

describe('快照测试', () => {
  it('默认', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <List>
            <ListItem>item</ListItem>
          </List>
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('设置属性', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <List>
            <ListItem insert disabledRipple disabled>
              item
            </ListItem>
          </List>
        </ThemeProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
