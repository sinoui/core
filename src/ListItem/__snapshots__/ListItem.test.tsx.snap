// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 设置属性 1`] = `
<ul
  className="sc-bdvvtL kCeUpj sinoui-list"
  onKeyDown={[Function]}
>
  <li
    className="sc-gsDKAQ bHyNFT sinoui-list-item sinoui-list-item--insert sinoui-list-item--disabled sinoui-list-item--disabled-ripple"
    disabled={true}
    tabIndex={0}
  >
    item
  </li>
</ul>
`;

exports[`快照测试 默认 1`] = `
<ul
  className="sc-bdvvtL kCeUpj sinoui-list"
  onKeyDown={[Function]}
>
  <li
    className="sc-gsDKAQ jMsphT sinoui-list-item"
    tabIndex={0}
  >
    item
  </li>
</ul>
`;
